File: packages/core/src/sdk/surface-protocol.ts
import type { ResolutionRequest, ResolutionResponse } from "../executor/index.js";
import { inspectLocalRunState } from "../runner-local/index.js";
import type { AuthResolver, Caller, ExecutionEvent, RunLocalSkillResult } from "../runner-local/index.js";

// Surface bridges let external hosts act as surfaces over the runx kernel.
// The host gets normalized run states; runx keeps ownership of execution,
// pause/resume, approvals, and receipts.
export interface SurfaceRunOptions {
  readonly skillPath: string;
  readonly inputs?: Readonly<Record<string, unknown>>;
  readonly answersPath?: string;
  readonly runner?: string;
  readonly receiptDir?: string;
  readonly runxHome?: string;
  readonly parentReceipt?: string;
  readonly contextFrom?: readonly string[];
  readonly caller?: Caller;
  readonly authResolver?: AuthResolver;
  readonly allowedSourceTypes?: readonly string[];
  readonly resumeFromRunId?: string;
}

export type SurfaceSkillExecutor = (
  options: SurfaceRunOptions & { readonly caller: Caller },
) => Promise<RunLocalSkillResult>;

export interface SurfaceBoundaryContext {
  readonly request: ResolutionRequest;
  readonly events: readonly ExecutionEvent[];
}

export type SurfaceBoundaryReply =
  | ResolutionResponse
  | {
      readonly actor?: "agent" | "human";
      readonly payload: unknown;
    }
  | boolean
  | string
  | number
  | Readonly<Record<string, unknown>>
  | undefined;

export type SurfaceBoundaryResolver = (
  context: SurfaceBoundaryContext,
) => Promise<SurfaceBoundaryReply> | SurfaceBoundaryReply;

export interface SurfaceBridgeOptions {
  readonly execute: SurfaceSkillExecutor;
  readonly inspect?: SurfaceStateInspector;
}

export interface SurfacePausedResult {
  readonly status: "paused";
  readonly skillName: string;
  readonly runId: string;
  readonly requests: readonly ResolutionRequest[];
  readonly stepIds?: readonly string[];
  readonly stepLabels?: readonly string[];
  readonly events: readonly ExecutionEvent[];
}

export interface SurfaceCompletedResult {
  readonly status: "completed";
  readonly skillName: string;
  readonly receiptId: string;
  readonly output: string;
  readonly events: readonly ExecutionEvent[];
}

export interface SurfaceFailedResult {
  readonly status: "failed";
  readonly skillName: string;
  readonly receiptId?: string;
  readonly error: string;
  readonly events: readonly ExecutionEvent[];
}

export interface SurfaceDeniedResult {
  readonly status: "denied";
  readonly skillName: string;
  readonly reasons: readonly string[];
  readonly receiptId?: string;
  readonly events: readonly ExecutionEvent[];
}

export type SurfaceRunResult =
  | SurfacePausedResult
  | SurfaceCompletedResult
  | SurfaceFailedResult
  | SurfaceDeniedResult;

export interface SurfaceRunVerification {
  readonly status: "verified" | "unverified" | "invalid";
  readonly reason?: string;
}

export interface SurfaceRunLineage {
  readonly kind: "rerun";
  readonly sourceRunId: string;
  readonly sourceReceiptId?: string;
}

export interface SurfaceRunApproval {
  readonly gateId?: string;
  readonly gateType?: string;
  readonly decision?: "approved" | "denied";
  readonly reason?: string;
}

export interface SurfaceInspectOptions {
  readonly receiptDir?: string;
  readonly runxHome?: string;
}

export type SurfaceStateInspector = (
  referenceId: string,
  options?: SurfaceInspectOptions,
) => Promise<SurfaceRunState>;

export interface SurfacePausedState {
  readonly status: "paused";
  readonly skillName: string;
  readonly runId: string;
  readonly requestedPath?: string;
  readonly resolvedPath?: string;
  readonly selectedRunner?: string;
  readonly requests: readonly ResolutionRequest[];
  readonly stepIds?: readonly string[];
  readonly stepLabels?: readonly string[];
  readonly lineage?: SurfaceRunLineage;
}

interface SurfaceTerminalState {
  readonly kind: "skill_execution" | "graph_execution";
  readonly skillName: string;
  readonly runId: string;
  readonly receiptId: string;
  readonly verification: SurfaceRunVerification;
  readonly sourceType?: string;
  readonly startedAt?: string;
  readonly completedAt?: string;
  readonly disposition?: string;
  readonly outcomeState?: string;
  readonly actors?: readonly string[];
  readonly artifactTypes?: readonly string[];
  readonly runnerProvider?: string;
  readonly approval?: SurfaceRunApproval;
  readonly lineage?: SurfaceRunLineage;
}

export interface SurfaceCompletedState extends SurfaceTerminalState {
  readonly status: "completed";
}

export interface SurfaceFailedState extends SurfaceTerminalState {
  readonly status: "failed";
}

export interface SurfaceDeniedState extends SurfaceTerminalState {
  readonly status: "denied";
}

export type SurfaceRunState =
  | SurfacePausedState
  | SurfaceCompletedState
  | SurfaceFailedState
  | SurfaceDeniedState;

export interface SurfaceBridge {
  readonly run: (
    options: SurfaceRunOptions & {
      readonly resolver?: SurfaceBoundaryResolver;
    },
  ) => Promise<SurfaceRunResult>;
  readonly resume: (
    runId: string,
    options: Omit<SurfaceRunOptions, "resumeFromRunId" | "skillPath"> & {
      readonly skillPath?: string;
      readonly resolver?: SurfaceBoundaryResolver;
    },
  ) => Promise<SurfaceRunResult>;
  readonly inspect: (
    referenceId: string,
    options?: SurfaceInspectOptions,
  ) => Promise<SurfaceRunState>;
}

export interface OpenAISurfaceResponse {
  readonly role: "tool";
  readonly content: readonly [{ readonly type: "text"; readonly text: string }];
  readonly structuredContent: {
    readonly runx: SurfaceRunResult;
  };
}

export interface AnthropicSurfaceResponse {
  readonly content: readonly [{ readonly type: "text"; readonly text: string }];
  readonly metadata: {
    readonly runx: SurfaceRunResult;
  };
}

export interface VercelAiSurfaceResponse {
  readonly messages: readonly [{ readonly role: "assistant"; readonly content: string }];
  readonly data: {
    readonly runx: SurfaceRunResult;
  };
}

export interface LangChainSurfaceResponse {
  readonly content: string;
  readonly additional_kwargs: {
    readonly runx: SurfaceRunResult;
  };
}

export interface CrewAiSurfaceResponse {
  readonly raw: string;
  readonly json_dict: {
    readonly runx: SurfaceRunResult;
  };
}

export interface ProviderSurfaceAdapter<TResponse> {
  readonly run: (
    options: SurfaceRunOptions & {
      readonly resolver?: SurfaceBoundaryResolver;
    },
  ) => Promise<TResponse>;
  readonly resume: (
    runId: string,
    options: Omit<SurfaceRunOptions, "resumeFromRunId" | "skillPath"> & {
      readonly skillPath?: string;
      readonly resolver?: SurfaceBoundaryResolver;
    },
  ) => Promise<TResponse>;
}

export function createSurfaceBridge(options: SurfaceBridgeOptions): SurfaceBridge {
  const bridge: SurfaceBridge = {
    run: async (runOptions) => {
      const events: ExecutionEvent[] = [];
      const caller: Caller = {
        report: async (event) => {
          events.push(event);
          await runOptions.caller?.report(event);
        },
        resolve: async (request) => {
          const resolved = normalizeSurfaceReply(await runOptions.resolver?.({ request, events }), request);
          return resolved ?? await runOptions.caller?.resolve(request);
        },
      };

      const result = await options.execute({
        ...runOptions,
        caller,
      });
      return normalizeRunResult(result, events);
    },
    resume: async (runId, runOptions) => {
      const skillPath = runOptions.skillPath ?? await resolveSurfaceResumeSkillPath(runId, runOptions, options.inspect);
      return await bridge.run({
        ...runOptions,
        skillPath,
        resumeFromRunId: runId,
      });
    },
    inspect: async (referenceId, inspectOptions) => {
      if (!options.inspect) {
        throw new Error("This surface bridge does not support inspect().");
      }
      return await options.inspect(referenceId, inspectOptions);
    },
  };
  return bridge;
}

export function createOpenAiSurfaceAdapter(bridge: SurfaceBridge): ProviderSurfaceAdapter<OpenAISurfaceResponse> {
  return {
    run: async (options) => toOpenAiResponse(await bridge.run(options)),
    resume: async (runId, options) => toOpenAiResponse(await bridge.resume(runId, options)),
  };
}

export function createAnthropicSurfaceAdapter(bridge: SurfaceBridge): ProviderSurfaceAdapter<AnthropicSurfaceResponse> {
  return {
    run: async (options) => toAnthropicResponse(await bridge.run(options)),
    resume: async (runId, options) => toAnthropicResponse(await bridge.resume(runId, options)),
  };
}

export function createVercelAiSurfaceAdapter(bridge: SurfaceBridge): ProviderSurfaceAdapter<VercelAiSurfaceResponse> {
  return {
    run: async (options) => toVercelAiResponse(await bridge.run(options)),
    resume: async (runId, options) => toVercelAiResponse(await bridge.resume(runId, options)),
  };
}

export function createLangChainSurfaceAdapter(bridge: SurfaceBridge): ProviderSurfaceAdapter<LangChainSurfaceResponse> {
  return {
    run: async (options) => toLangChainResponse(await bridge.run(options)),
    resume: async (runId, options) => toLangChainResponse(await bridge.resume(runId, options)),
  };
}

export function createCrewAiSurfaceAdapter(bridge: SurfaceBridge): ProviderSurfaceAdapter<CrewAiSurfaceResponse> {
  return {
    run: async (options) => toCrewAiResponse(await bridge.run(options)),
    resume: async (runId, options) => toCrewAiResponse(await bridge.resume(runId, options)),
  };
}

function normalizeSurfaceReply(
  reply: SurfaceBoundaryReply,
  request: ResolutionRequest,
): ResolutionResponse | undefined {
  if (reply === undefined) {
    return undefined;
  }
  if (isResolutionResponse(reply)) {
    return reply;
  }
  if (typeof reply === "object" && reply !== null && "payload" in reply) {
    const candidate = reply as { readonly actor?: "agent" | "human"; readonly payload: unknown };
    return {
      actor: candidate.actor ?? defaultActorForRequest(request),
      payload: candidate.payload,
    };
  }
  if (typeof reply === "boolean" && request.kind === "approval") {
    return { actor: "human", payload: reply };
  }
  return {
    actor: defaultActorForRequest(request),
    payload: reply,
  };
}

function defaultActorForRequest(request: ResolutionRequest): "agent" | "human" {
  return request.kind === "cognitive_work" ? "agent" : "human";
}

function isResolutionResponse(value: unknown): value is ResolutionResponse {
  return typeof value === "object"
    && value !== null
    && "actor" in value
    && "payload" in value
    && (((value as { readonly actor?: unknown }).actor === "agent") || ((value as { readonly actor?: unknown }).actor === "human"));
}

function normalizeRunResult(result: RunLocalSkillResult, events: readonly ExecutionEvent[]): SurfaceRunResult {
  if (result.status === "needs_resolution") {
    return {
      status: "paused",
      skillName: result.skill.name,
      runId: result.runId,
      requests: result.requests,
      stepIds: result.stepIds,
      stepLabels: result.stepLabels,
      events,
    };
  }
  if (result.status === "policy_denied") {
    return {
      status: "denied",
      skillName: result.skill.name,
      reasons: result.reasons,
      receiptId: result.receipt?.id,
      events,
    };
  }
  if (result.status === "success") {
    return {
      status: "completed",
      skillName: result.skill.name,
      receiptId: result.receipt.id,
      output: result.execution.stdout,
      events,
    };
  }
  return {
    status: "failed",
    skillName: result.skill.name,
    receiptId: result.receipt.id,
    error: result.execution.errorMessage ?? (result.execution.stderr || result.execution.stdout),
    events,
  };
}

export async function inspectLocalSurfaceState(
  referenceId: string,
  options: SurfaceInspectOptions & {
    readonly env?: NodeJS.ProcessEnv;
  } = {},
): Promise<SurfaceRunState> {
  const inspected = await inspectLocalRunState({
    referenceId,
    receiptDir: options.receiptDir,
    runxHome: options.runxHome,
    env: options.env,
  });
  if (inspected.status === "paused") {
    return {
      status: "paused",
      skillName: inspected.pending.skillName ?? deriveSkillNameFromPending(inspected.pending),
      runId: inspected.runId,
      requestedPath: inspected.pending.skillPath,
      resolvedPath: inspected.pending.resolvedSkillPath,
      selectedRunner: inspected.pending.selectedRunner,
      requests: inspected.pending.requests ?? [],
      stepIds: inspected.pending.stepIds,
      stepLabels: inspected.pending.stepLabels,
      lineage: inspected.pending.lineage,
    };
  }

  const status = inspectStatus(inspected.summary);
  return {
    status,
    kind: inspected.summary.kind,
    skillName: inspected.summary.name,
    runId: inspected.runId,
    receiptId: inspected.receipt.id,
    verification: inspected.verification,
    sourceType: inspected.summary.sourceType,
    startedAt: inspected.summary.startedAt,
    completedAt: inspected.summary.completedAt,
    disposition: inspected.summary.disposition,
    outcomeState: inspected.summary.outcomeState,
    actors: inspected.summary.actors,
    artifactTypes: inspected.summary.artifactTypes,
    runnerProvider: inspected.summary.runnerProvider,
    approval: inspected.summary.approval,
    lineage: inspected.summary.lineage,
  };
}

function inspectStatus(summary: {
  readonly status: string;
  readonly disposition?: string;
}): SurfaceCompletedState["status"] | SurfaceFailedState["status"] | SurfaceDeniedState["status"] {
  if (summary.disposition === "policy_denied") {
    return "denied";
  }
  return summary.status === "success" ? "completed" : "failed";
}

function deriveSkillNameFromPending(pending: {
  readonly skillName?: string;
  readonly skillPath?: string;
  readonly resolvedSkillPath?: string;
}): string {
  if (pending.skillName && pending.skillName.trim().length > 0) {
    return pending.skillName;
  }
  const candidate = pending.resolvedSkillPath ?? pending.skillPath;
  if (!candidate) {
    return "unknown";
  }
  const normalized = candidate.replace(/\\/g, "/");
  const trimmed = normalized.endsWith("/SKILL.md")
    ? normalized.slice(0, -"/SKILL.md".length)
    : normalized.endsWith(".md")
      ? normalized.slice(0, -".md".length)
      : normalized;
  const segments = trimmed.split("/").filter((segment) => segment.length > 0);
  return segments[segments.length - 1] ?? "unknown";
}

async function resolveSurfaceResumeSkillPath(
  runId: string,
  options: SurfaceInspectOptions & {
    readonly skillPath?: string;
  },
  inspect?: SurfaceStateInspector,
): Promise<string> {
  if (options.skillPath) {
    return options.skillPath;
  }
  if (!inspect) {
    throw new Error(`Run '${runId}' cannot be resumed because this surface bridge cannot resolve pending skill paths.`);
  }
  const state = await inspect(runId, options);
  if (state.status !== "paused") {
    throw new Error(`Run '${runId}' is not paused and cannot be resumed.`);
  }
  const skillPath = state.requestedPath ?? state.resolvedPath;
  if (!skillPath) {
    throw new Error(`Run '${runId}' cannot be resumed because no pending skill path was recorded.`);
  }
  return skillPath;
}

function toOpenAiResponse(result: SurfaceRunResult): OpenAISurfaceResponse {
  return {
    role: "tool",
    content: [{ type: "text", text: summarizeSurfaceResult(result) }],
    structuredContent: { runx: result },
  };
}

function toAnthropicResponse(result: SurfaceRunResult): AnthropicSurfaceResponse {
  return {
    content: [{ type: "text", text: summarizeSurfaceResult(result) }],
    metadata: { runx: result },
  };
}

function toVercelAiResponse(result: SurfaceRunResult): VercelAiSurfaceResponse {
  return {
    messages: [{ role: "assistant", content: summarizeSurfaceResult(result) }],
    data: { runx: result },
  };
}

function toLangChainResponse(result: SurfaceRunResult): LangChainSurfaceResponse {
  return {
    content: summarizeSurfaceResult(result),
    additional_kwargs: { runx: result },
  };
}

function toCrewAiResponse(result: SurfaceRunResult): CrewAiSurfaceResponse {
  return {
    raw: summarizeSurfaceResult(result),
    json_dict: { runx: result },
  };
}

function summarizeSurfaceResult(result: SurfaceRunResult): string {
  switch (result.status) {
    case "completed":
      return `${result.skillName} completed. Inspect receipt ${result.receiptId}.`;
    case "paused":
      return `${result.skillName} paused at ${result.runId}. Resume after resolving ${result.requests.length} request(s).`;
    case "denied":
      return `${result.skillName} was denied by policy.`;
    case "failed":
      return `${result.skillName} failed. Inspect receipt ${result.receiptId ?? "n/a"}.`;
  }
}

File: packages/sdk-python/runx/surface_protocol.py
from __future__ import annotations

from dataclasses import dataclass
from typing import Any, Callable, Mapping, Sequence

from . import RunxClient


@dataclass(frozen=True)
class SurfaceBoundaryContext:
    request: Mapping[str, Any]
    events: tuple[Mapping[str, Any], ...] = ()


SurfaceBoundaryResolver = Callable[[SurfaceBoundaryContext], Any | None]


@dataclass(frozen=True)
class SurfacePausedResult:
    status: str
    skill_name: str
    run_id: str
    requests: tuple[Mapping[str, Any], ...]
    step_ids: tuple[str, ...] = ()
    step_labels: tuple[str, ...] = ()
    events: tuple[Mapping[str, Any], ...] = ()


@dataclass(frozen=True)
class SurfaceCompletedResult:
    status: str
    skill_name: str
    receipt_id: str
    output: str
    events: tuple[Mapping[str, Any], ...] = ()


@dataclass(frozen=True)
class SurfaceFailedResult:
    status: str
    skill_name: str
    error: str
    receipt_id: str | None = None
    events: tuple[Mapping[str, Any], ...] = ()


@dataclass(frozen=True)
class SurfaceDeniedResult:
    status: str
    skill_name: str
    reasons: tuple[str, ...]
    receipt_id: str | None = None
    events: tuple[Mapping[str, Any], ...] = ()


SurfaceRunResult = (
    SurfacePausedResult
    | SurfaceCompletedResult
    | SurfaceFailedResult
    | SurfaceDeniedResult
)


@dataclass(frozen=True)
class SurfacePausedState:
    status: str
    skill_name: str
    run_id: str
    requested_path: str | None = None
    resolved_path: str | None = None
    selected_runner: str | None = None
    requests: tuple[Mapping[str, Any], ...] = ()
    step_ids: tuple[str, ...] = ()
    step_labels: tuple[str, ...] = ()
    lineage: Mapping[str, Any] | None = None


@dataclass(frozen=True)
class SurfaceTerminalState:
    status: str
    kind: str
    skill_name: str
    run_id: str
    receipt_id: str
    verification: Mapping[str, Any]
    source_type: str | None = None
    started_at: str | None = None
    completed_at: str | None = None
    disposition: str | None = None
    outcome_state: str | None = None
    actors: tuple[str, ...] = ()
    artifact_types: tuple[str, ...] = ()
    runner_provider: str | None = None
    approval: Mapping[str, Any] | None = None
    lineage: Mapping[str, Any] | None = None


SurfaceRunState = SurfacePausedState | SurfaceTerminalState


class SurfaceBridge:
    def __init__(self, client: RunxClient) -> None:
        self.client = client

    def run(
        self,
        skill_path: str,
        inputs: Mapping[str, Any] | None = None,
        resolver: SurfaceBoundaryResolver | None = None,
    ) -> SurfaceRunResult:
        initial = self.client.surface_run(skill_path, inputs=inputs)
        return self._drive(initial, resolver=resolver)

    def resume(
        self,
        run_id: str,
        resolver: SurfaceBoundaryResolver | None = None,
    ) -> SurfaceRunResult:
        initial = self.client.surface_resume(run_id)
        return self._drive(initial, resolver=resolver)

    def inspect(self, reference_id: str) -> SurfaceRunState:
        return normalize_surface_state(self.client.surface_inspect(reference_id))

    def _drive(
        self,
        payload: Mapping[str, Any],
        resolver: SurfaceBoundaryResolver | None,
    ) -> SurfaceRunResult:
        current = dict(payload)
        while True:
            result = normalize_surface_result(current)
            if not isinstance(result, SurfacePausedResult):
                return result
            if resolver is None:
                return result

            responses: list[dict[str, Any]] = []
            for request in result.requests:
                reply = resolver(SurfaceBoundaryContext(request=request, events=result.events))
                normalized = _normalize_resolution_reply(request, reply)
                if normalized is None:
                    continue
                responses.append(
                    {
                        "requestId": str(request.get("id") or ""),
                        "actor": normalized["actor"],
                        "payload": normalized["payload"],
                    }
                )

            if not responses:
                return result

            current = self.client.surface_resume(result.run_id, responses=responses)


class ProviderSurfaceAdapter:
    def __init__(self, bridge: SurfaceBridge, formatter: Callable[[SurfaceRunResult], Mapping[str, Any]]) -> None:
        self.bridge = bridge
        self.formatter = formatter

    def run(
        self,
        skill_path: str,
        inputs: Mapping[str, Any] | None = None,
        resolver: SurfaceBoundaryResolver | None = None,
    ) -> Mapping[str, Any]:
        return self.formatter(self.bridge.run(skill_path, inputs=inputs, resolver=resolver))

    def resume(
        self,
        run_id: str,
        resolver: SurfaceBoundaryResolver | None = None,
    ) -> Mapping[str, Any]:
        return self.formatter(self.bridge.resume(run_id, resolver=resolver))


def create_surface_bridge(client: RunxClient) -> SurfaceBridge:
    return SurfaceBridge(client)


def create_openai_surface_adapter(bridge: SurfaceBridge) -> ProviderSurfaceAdapter:
    return ProviderSurfaceAdapter(bridge, _to_openai_response)


def create_anthropic_surface_adapter(bridge: SurfaceBridge) -> ProviderSurfaceAdapter:
    return ProviderSurfaceAdapter(bridge, _to_anthropic_response)


def create_vercel_ai_surface_adapter(bridge: SurfaceBridge) -> ProviderSurfaceAdapter:
    return ProviderSurfaceAdapter(bridge, _to_vercel_response)


def create_langchain_surface_adapter(bridge: SurfaceBridge) -> ProviderSurfaceAdapter:
    return ProviderSurfaceAdapter(bridge, _to_langchain_response)


def create_crewai_surface_adapter(bridge: SurfaceBridge) -> ProviderSurfaceAdapter:
    return ProviderSurfaceAdapter(bridge, _to_crewai_response)


def normalize_surface_result(payload: Mapping[str, Any]) -> SurfaceRunResult:
    if _is_canonical_surface_result(payload):
        return _normalize_canonical_surface_result(payload)

    status = str(payload.get("status") or "")
    skill = payload.get("skill")
    skill_name = str(skill.get("name")) if isinstance(skill, Mapping) else str(skill or "")
    if status == "needs_resolution":
        return SurfacePausedResult(
            status="paused",
            skill_name=skill_name,
            run_id=str(payload.get("run_id") or ""),
            requests=tuple(payload.get("requests") or ()),
            step_ids=tuple(str(item) for item in payload.get("step_ids") or ()),
            step_labels=tuple(str(item) for item in payload.get("step_labels") or ()),
        )
    if status == "policy_denied":
        reasons = payload.get("reasons") or ()
        receipt = payload.get("receipt") or {}
        return SurfaceDeniedResult(
            status="denied",
            skill_name=skill_name,
            reasons=tuple(str(item) for item in reasons),
            receipt_id=_nested_str(receipt, "id"),
        )
    if status == "success":
        execution = payload.get("execution") or {}
        receipt = payload.get("receipt") or {}
        return SurfaceCompletedResult(
            status="completed",
            skill_name=skill_name,
            receipt_id=str(receipt.get("id") or ""),
            output=str(execution.get("stdout") or ""),
        )
    execution = payload.get("execution") or {}
    receipt = payload.get("receipt") or {}
    error = str(execution.get("errorMessage") or execution.get("stderr") or execution.get("stdout") or "")
    return SurfaceFailedResult(
        status="failed",
        skill_name=skill_name,
        error=error,
        receipt_id=_nested_str(receipt, "id"),
    )


def normalize_surface_state(payload: Mapping[str, Any]) -> SurfaceRunState:
    status = str(payload.get("status") or "")
    if status == "paused":
        return SurfacePausedState(
            status="paused",
            skill_name=str(payload.get("skillName") or ""),
            run_id=str(payload.get("runId") or ""),
            requested_path=_optional_str(payload.get("requestedPath")),
            resolved_path=_optional_str(payload.get("resolvedPath")),
            selected_runner=_optional_str(payload.get("selectedRunner")),
            requests=tuple(payload.get("requests") or ()),
            step_ids=tuple(str(item) for item in payload.get("stepIds") or ()),
            step_labels=tuple(str(item) for item in payload.get("stepLabels") or ()),
            lineage=payload.get("lineage") if isinstance(payload.get("lineage"), Mapping) else None,
        )
    return SurfaceTerminalState(
        status=status,
        kind=str(payload.get("kind") or ""),
        skill_name=str(payload.get("skillName") or ""),
        run_id=str(payload.get("runId") or ""),
        receipt_id=str(payload.get("receiptId") or ""),
        verification=dict(payload.get("verification") or {}),
        source_type=_optional_str(payload.get("sourceType")),
        started_at=_optional_str(payload.get("startedAt")),
        completed_at=_optional_str(payload.get("completedAt")),
        disposition=_optional_str(payload.get("disposition")),
        outcome_state=_optional_str(payload.get("outcomeState")),
        actors=tuple(str(item) for item in payload.get("actors") or ()),
        artifact_types=tuple(str(item) for item in payload.get("artifactTypes") or ()),
        runner_provider=_optional_str(payload.get("runnerProvider")),
        approval=payload.get("approval") if isinstance(payload.get("approval"), Mapping) else None,
        lineage=payload.get("lineage") if isinstance(payload.get("lineage"), Mapping) else None,
    )


def _normalize_resolution_reply(
    request: Mapping[str, Any],
    reply: Any | None,
) -> Mapping[str, Any] | None:
    if reply is None:
        return None
    if isinstance(reply, Mapping) and "actor" in reply and "payload" in reply:
        return {
            "actor": str(reply.get("actor") or _default_actor_for_request(request)),
            "payload": reply.get("payload"),
        }
    if isinstance(reply, Mapping) and "payload" in reply:
        return {
            "actor": str(reply.get("actor") or _default_actor_for_request(request)),
            "payload": reply.get("payload"),
        }
    if isinstance(reply, bool) and request.get("kind") == "approval":
        return {"actor": "human", "payload": reply}
    return {
        "actor": _default_actor_for_request(request),
        "payload": reply,
    }


def _default_actor_for_request(request: Mapping[str, Any]) -> str:
    return "agent" if request.get("kind") == "cognitive_work" else "human"


def _summary(result: SurfaceRunResult) -> str:
    if isinstance(result, SurfaceCompletedResult):
        return f"{result.skill_name} completed. Inspect receipt {result.receipt_id}."
    if isinstance(result, SurfacePausedResult):
        return f"{result.skill_name} paused at {result.run_id}. Resume after resolving {len(result.requests)} request(s)."
    if isinstance(result, SurfaceDeniedResult):
        return f"{result.skill_name} was denied by policy."
    return f"{result.skill_name} failed. Inspect receipt {result.receipt_id or 'n/a'}."


def _is_canonical_surface_result(payload: Mapping[str, Any]) -> bool:
    return isinstance(payload.get("skillName"), str) and str(payload.get("status") or "") in {
        "paused",
        "completed",
        "failed",
        "denied",
    }


def _normalize_canonical_surface_result(payload: Mapping[str, Any]) -> SurfaceRunResult:
    status = str(payload.get("status") or "")
    if status == "paused":
        return SurfacePausedResult(
            status="paused",
            skill_name=str(payload.get("skillName") or ""),
            run_id=str(payload.get("runId") or ""),
            requests=tuple(payload.get("requests") or ()),
            step_ids=tuple(str(item) for item in payload.get("stepIds") or ()),
            step_labels=tuple(str(item) for item in payload.get("stepLabels") or ()),
            events=tuple(payload.get("events") or ()),
        )
    if status == "completed":
        return SurfaceCompletedResult(
            status="completed",
            skill_name=str(payload.get("skillName") or ""),
            receipt_id=str(payload.get("receiptId") or ""),
            output=str(payload.get("output") or ""),
            events=tuple(payload.get("events") or ()),
        )
    if status == "denied":
        return SurfaceDeniedResult(
            status="denied",
            skill_name=str(payload.get("skillName") or ""),
            reasons=tuple(str(item) for item in payload.get("reasons") or ()),
            receipt_id=_optional_str(payload.get("receiptId")),
            events=tuple(payload.get("events") or ()),
        )
    return SurfaceFailedResult(
        status="failed",
        skill_name=str(payload.get("skillName") or ""),
        error=str(payload.get("error") or ""),
        receipt_id=_optional_str(payload.get("receiptId")),
        events=tuple(payload.get("events") or ()),
    )


def _to_openai_response(result: SurfaceRunResult) -> Mapping[str, Any]:
    return {
        "role": "tool",
        "content": [{"type": "text", "text": _summary(result)}],
        "structuredContent": {"runx": _result_to_dict(result)},
    }


def _to_anthropic_response(result: SurfaceRunResult) -> Mapping[str, Any]:
    return {
        "content": [{"type": "text", "text": _summary(result)}],
        "metadata": {"runx": _result_to_dict(result)},
    }


def _to_vercel_response(result: SurfaceRunResult) -> Mapping[str, Any]:
    return {
        "messages": [{"role": "assistant", "content": _summary(result)}],
        "data": {"runx": _result_to_dict(result)},
    }


def _to_langchain_response(result: SurfaceRunResult) -> Mapping[str, Any]:
    return {
        "content": _summary(result),
        "additional_kwargs": {"runx": _result_to_dict(result)},
    }


def _to_crewai_response(result: SurfaceRunResult) -> Mapping[str, Any]:
    return {
        "raw": _summary(result),
        "json_dict": {"runx": _result_to_dict(result)},
    }


def _result_to_dict(result: SurfaceRunResult) -> Mapping[str, Any]:
    if isinstance(result, SurfacePausedResult):
        return {
            "status": result.status,
            "skillName": result.skill_name,
            "runId": result.run_id,
            "requests": list(result.requests),
            "stepIds": list(result.step_ids),
            "stepLabels": list(result.step_labels),
            "events": list(result.events),
        }
    if isinstance(result, SurfaceCompletedResult):
        return {
            "status": result.status,
            "skillName": result.skill_name,
            "receiptId": result.receipt_id,
            "output": result.output,
            "events": list(result.events),
        }
    if isinstance(result, SurfaceDeniedResult):
        return {
            "status": result.status,
            "skillName": result.skill_name,
            "reasons": list(result.reasons),
            "receiptId": result.receipt_id,
            "events": list(result.events),
        }
    return {
        "status": result.status,
        "skillName": result.skill_name,
        "error": result.error,
        "receiptId": result.receipt_id,
        "events": list(result.events),
    }


def _nested_str(payload: Mapping[str, Any], key: str) -> str | None:
    value = payload.get(key)
    return None if value is None else str(value)


def _optional_str(value: Any) -> str | None:
    return None if value is None else str(value)
